src/interpreter/example-validator.ts

```diff
@@ -0,0 +1,482 @@
+
+    return this.errors.length > 0
+      ? { pass: false, errors: this.errors, warnings: this.warnings }
+      : { pass: true, warnings: this.warnings };
+  }
+
+  visit(node: ComlinkLiteralNode): boolean;
+  visit(node: ProfileASTNode): void;
+  visit(node: ProfileASTNode): boolean | void {
+    debug('Visiting node:', node.kind);
+
+    switch (node.kind) {
+      case 'ProfileDocument':
+        return this.visitProfileDocumentNode(node);
+      case 'ProfileHeader':
+        return this.visitProfileHeaderNode(node);
+      case 'UseCaseDefinition':
+        return this.visitUseCaseDefinitionNode(node);
+      case 'UseCaseSlotDefinition':
+        return this.visitUseCaseSlotDefinition(node);
+      case 'UseCaseExample':
+        return this.visitUseCaseExampleNode(node);
+      case 'ComlinkPrimitiveLiteral':
+        return this.visitComlinkPrimitiveLiteralNode(node);
+      case 'ComlinkListLiteral':
+        return this.visitComlinkListLiteralNode(node);
+      case 'ComlinkObjectLiteral':
+        return this.visitComlinkObjectLiteralNode(node);
+      // UNUSED
+      case 'FieldDefinition':
+        return this.visitFieldDefinitionNode(node);
+      case 'ModelTypeName':
+        return this.visitModelTypeNameNode(node);
+      case 'NamedFieldDefinition':
+        return this.visitNamedFieldDefinitionNode(node);
+      case 'NamedModelDefinition':
+        return this.visitNamedModelDefinitionNode(node);
+      case 'NonNullDefinition':
+        return this.visitNonNullDefinitionNode(node);
+      case 'UnionDefinition':
+        return this.visitUnionDefinitionNode(node);
+      case 'ObjectDefinition':
+        return this.visitObjectDefinitionNode(node);
+      case 'ListDefinition':
+        return this.visitListDefinitionNode(node);
+      case 'EnumDefinition':
+        return this.visitEnumDefinitionNode(node);
+      case 'EnumValue':
+        return this.visitEnumValueNode(node);
+      case 'PrimitiveTypeName':
+        return this.visitPrimitiveTypeNameNode(node);
+      default:
+        assertUnreachable(node);
+    }
+  }
+
+  visitProfileDocumentNode(node: ProfileDocumentNode): void {
+    const useCases = node.definitions.filter(isUseCaseDefinitionNode);
+
+    for (const useCase of useCases) {
+      this.visit(useCase);
+    }
+  }
+
+  visitProfileHeaderNode(_node: ProfileHeaderNode): void {
+    throw new Error('not implemented');
+  }
+
+  visitUseCaseDefinitionNode(node: UseCaseDefinitionNode): void {
+    this.currentUseCase = node.useCaseName;
+
+    if (node.examples) {
+      for (const example of node.examples) {
+        this.visit(example.value);
+      }
+    }
+
+    this.currentUseCase = undefined;
+  }
+
+  visitUseCaseSlotDefinition(
+    node: UseCaseSlotDefinitionNode<ProfileASTNode>
+  ): void {
+    this.visit(node.value);
+  }
+
+  visitUseCaseExampleNode(node: UseCaseExampleNode): void {
+    if (node.input) {
+      this.currentStructure = this.profileOutput?.usecases.find(
```

Comment:
You can do `find` before ifs 🤷

Reply:
true 😄

Reply:
And maybe you can do something with this.slotType = undefined and this.currentStructure = undefined